plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,134 @@
+     */
+    public static final String QUERY_GROUP_COUNT_SETTING_NAME = "node.query_group.max_count";
+    /**
+     * default max queryGroup count on any node at any given point in time
+     */
+    private static final int DEFAULT_MAX_QUERY_GROUP_COUNT_VALUE = 100;
+    /**
+     * min queryGroup count on any node at any given point in time
+     */
+    private static final int MIN_QUERY_GROUP_COUNT_VALUE = 1;
+    /**
+     *  max QueryGroup count setting
+     */
+    public static final Setting<Integer> MAX_QUERY_GROUP_COUNT = Setting.intSetting(
+        QUERY_GROUP_COUNT_SETTING_NAME,
+        DEFAULT_MAX_QUERY_GROUP_COUNT_VALUE,
+        0,
+        QueryGroupPersistenceService::validateMaxQueryGroupCount,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private final ClusterService clusterService;
+    private volatile int maxQueryGroupCount;
+
+    /**
+     * Constructor for QueryGroupPersistenceService
+     *
+     * @param clusterService {@link ClusterService} - The cluster service to be used by QueryGroupPersistenceService
+     * @param settings {@link Settings} - The settings to be used by QueryGroupPersistenceService
+     * @param clusterSettings {@link ClusterSettings} - The cluster settings to be used by QueryGroupPersistenceService
+     */
+    @Inject
+    public QueryGroupPersistenceService(
+        final ClusterService clusterService,
+        final Settings settings,
+        final ClusterSettings clusterSettings
+    ) {
+        this.clusterService = clusterService;
+        setMaxQueryGroupCount(MAX_QUERY_GROUP_COUNT.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(MAX_QUERY_GROUP_COUNT, this::setMaxQueryGroupCount);
+    }
+
+    /**
+     * Set maxQueryGroupCount to be newMaxQueryGroupCount
+     * @param newMaxQueryGroupCount - the max number of QueryGroup allowed
+     */
+    public void setMaxQueryGroupCount(int newMaxQueryGroupCount) {
+        validateMaxQueryGroupCount(newMaxQueryGroupCount);
+        this.maxQueryGroupCount = newMaxQueryGroupCount;
+    }
+
+    /**
+     * Validator for maxQueryGroupCount
+     * @param maxQueryGroupCount - the maxQueryGroupCount number to be verified
+     */
+    private static void validateMaxQueryGroupCount(int maxQueryGroupCount) {
+        if (maxQueryGroupCount > DEFAULT_MAX_QUERY_GROUP_COUNT_VALUE || maxQueryGroupCount < MIN_QUERY_GROUP_COUNT_VALUE) {
+            throw new IllegalArgumentException(QUERY_GROUP_COUNT_SETTING_NAME + " should be in range [1-100].");
+        }
+    }
+
+    /**
+     * Get the QueryGroup with the specified name and generate response
+     * @param name - the QueryGroup name we are getting
+     * @param listener - ActionListener for GetQueryGroupResponse
+     */
+    public void getFromClusterStateMetadata(String name, ActionListener<GetQueryGroupResponse> listener) {
+        List<QueryGroup> resultGroups = getQueryGroupsFromClusterState(name, clusterService.state());
```

Comment:
This is the cluster service on request receiving the request which might be lagging the cluster state version. We should fetch this information on master node to ensure we get the query group from latest information. This also breaks the eventually consistency constraint, since the data nodes might be temporarily inconsistent which can reflect in the query group information that we share with the client. First invocation might show a query group, the next might not show query group